lib/core-integration/src/Test/Integration/Scenario/API/Shelley/SharedWallets.hs

```diff
@@ -0,0 +1,607 @@
+        expectErrorMessage (errMsg403KeyAlreadyPresent (toText Delegation)) rPatch4
+
+        let payloadPatch5 = Json [json| {
+                "cosigner#7": #{accXPub0}
+                } |]
+        rPatch5 <- patchSharedWallet ctx wal Payment payloadPatch5
+        expectResponseCode HTTP.status403 rPatch5
+        expectErrorMessage (errMsg403NoSuchCosigner (toText Payment) 7) rPatch5
+
+    it "SHARED_WALLETS_PATCH_07 - Cannot update cosigner key in a pending shared wallet having the shared wallet's account key" $ \ctx -> runResourceT $ do
+        let payloadCreate = Json [json| {
+                "name": "Shared Wallet",
+                "account_public_key": #{accXPubTxt0},
+                "account_index": "30H",
+                "payment_script_template":
+                    { "cosigners":
+                        { "cosigner#0": #{accXPubTxt0} },
+                      "template":
+                          { "all":
+                             [ "cosigner#0",
+                               "cosigner#1",
+                               { "active_from": 120 }
+                             ]
+                          }
+                    }
+                } |]
+        rPost <- postSharedWallet ctx Default payloadCreate
+        expectResponseCode HTTP.status201 rPost
+        let wal@(ApiSharedWallet (Left _pendingWal)) = getFromResponse id rPost
+
+        let payloadPatch = Json [json| {
+                "cosigner#0": #{accXPub1}
+                } |]
+        rPatch <- patchSharedWallet ctx wal Payment payloadPatch
+        expectResponseCode HTTP.status403 rPatch
+        expectErrorMessage errMsg403CannotUpdateThisCosigner rPatch
+
+    it "SHARED_WALLETS_PATCH_08 - Active wallet needs shared wallet' account key in template" $ \ctx -> runResourceT $ do
+        let payloadCreate = Json [json| {
+                "name": "Shared Wallet",
+                "account_public_key": #{accXPubTxt0},
+                "account_index": "30H",
+                "payment_script_template":
+                    { "cosigners":
+                        { "cosigner#0": #{accXPubTxt1} },
+                      "template":
+                          { "all":
+                             [ "cosigner#0",
+                               { "active_from": 120 }
+                             ]
+                          }
+                    }
+                } |]
+        rPost <- postSharedWallet ctx Default payloadCreate
+        expectResponseCode HTTP.status201 rPost
+        let wal@(ApiSharedWallet (Left _pendingWal)) = getFromResponse id rPost
+
+        let payloadPatch = Json [json| {
+                "cosigner#0": #{accXPub0}
+                } |]
+
+        rPatch <- patchSharedWallet ctx wal Payment payloadPatch
+        expectResponseCode HTTP.status201 rPost
+        let (ApiSharedWallet walletConstructor) = getFromResponse id rPatch
+        liftIO $ isRight walletConstructor `shouldBe` True
```

Comment:
So we have one cosigner payment script template with accXPub different than the one in shared wallet. So although we have all cosigner keys collected we are in pending state (isLeft is True). Only after we patch the cosigner with the accXPub equal to the one in shared wallet we have transition from pending to active - manifested here that isRight is True. This is an illustrative example that validation takes care of the condition that the shared wallet key is PRESENT in cosigner keys.

Reply:
Ho. But that is wrong isn't it? I didn't notice that at first but, after the creation, that wallet isn't pending; it's fully complete and active because all the cosigners are provided at creation time. So that patch is actually pointless (and I'd even expect an error 'wallet already active' in that case)

Reply:
no it is in pending after posting

Reply:
```
568         let wal@(ApiSharedWallet (Left _pendingWal)) = getFromResponse id rPost
```
And it should be so (to my best knowledge) as we have : 
```
552                 "name": "Shared Wallet",                                                                                                                                                                       
553                 "account_public_key": #{accXPubTxt0},                                                                                                                                                          
554                 "account_index": "30H",                                                                                                                                                                        
555                 "payment_script_template":                                                                                                                                                                     
556                     { "cosigners":                                                                                                                                                                             
557                         { "cosigner#0": #{accXPubTxt1} },                                                                                                                                                      
558                       "template":                                                                                                                                                                              
559                           { "all":                                                                                                                                                                             
560                              [ "cosigner#0",                                                                                                                                                                   
561                                { "active_from": 120 }                                                                                                                                                          
562                              ]                                                                                                                                                                                 
563                           }                                                                                                                                                                                    
564                     }                           
```
account public key of shared wallet is different than the one in cosigner. And after updating to the accXPub of the shared wallet we are in acive state:
```
576         let (ApiSharedWallet walletConstructor) = getFromResponse id rPatch                                                                                                                                    
577         liftIO $ isRight walletConstructor `shouldBe` True 
```

Reply:
> account public key of shared wallet is different than the one in cosigner. 

This should not be possible! Creating such wallet should lead to an error because this totally breaks the invariant stated earlier: 

a shared wallet must always, at all time, contains its top-level account key as cosigner. 

That includes during creation. As a corollary, shared wallet with single cosigners never reach the pending state, they are either immediately valid or, invalid.

Reply:
ok, got it. so this invariant MUST be valid not ONLY for active, but also PENDING state -> I will make sure this is the case soon. Thanks